web/src/engine/websites/AsuraScans.ts

```diff
@@ -47,8 +58,40 @@ export default class extends DecoratableMangaScraper {
     }
 
     public override async FetchChapters(manga: Manga): Promise<Chapter[]> {
-        const chapters = await FetchCSS<HTMLAnchorElement>(new Request(new URL(manga.Identifier, this.URI)), 'div.scrollbar-thumb-themecolor a.block');
-        return chapters.map(chapter => new Chapter(this, manga, [manga.Identifier, 'chapter', chapter.pathname.match(/(\d+)+$/)[1]].join('/'), chapter.textContent.trim()));
+        const scripts = await FetchCSS<HTMLScriptElement>(new Request(new URL(`${manga.Identifier}`, this.URI)), 'script:not([src])');
+        const { chapters } = this.FindJSONObject<JSONChapters>(scripts, /is_early_access/, 'chapters');
+        return chapters.map(chapter => new Chapter(this, manga, [manga.Identifier, 'chapter', chapter.name.toString()].join('/'), `Chapter ${chapter.name}`));
     }
 
+    public override async FetchPages(chapter: Chapter): Promise<Page[]> {
+        const scripts = await FetchCSS<HTMLScriptElement>(new Request(new URL(`${chapter.Identifier}`, this.URI)), 'script:not([src])');
+        const { pages } = this.FindJSONObject<JSONPages>(scripts, /chapterName/, 'pages');
+        return pages
+            .filter(page => excludes.none(pattern => pattern.test(page.url)))
+            .map(page => new Page(this, chapter, new URL(page.url)));
+    }
+
+    private FindJSONObject<T>(scripts: HTMLScriptElement[], scriptRegex: RegExp, keyName: string, currentElement = undefined): T {
```

Comment:
Split responsibilities => first method for parsing scripts to objects, second method for searching key in objects

Reply:
You told me to use one single method in another plugin, now you want two 😢

Reply:
Don't know about the other comment, but in this specific case the parameters `scripts` and `scriptRegex` are only used once for parsing and become redundant (are used as flags for conditionally excluding part of the business logic in consecutive calls) which indicates a violation of the single responsibility principle.

It would be a different story if there was no recursive call.

Reply:
Found it in #821

The remark was related to combining the repetition of multiple calls.

```javascript
// somewhere
const tmp1 = A();
const res1 = B(tmp1);

// somewhere else
const tmp2 = A();
const res2 = B(tmp2);
```

Which was addressed by merging the business logic of `A()` and `B()` into a single replacement function, but could have also been addressed by grouping these calls into another method.
```javascript

C() {
  const tmp = A();
  return B(tmp);
}

// somewhere
const res1 = C();

// somewhere else
const res2 = C();
```

To be fair the final implementation of the single replacement function (which is also presented in this PR) was merged into `master` directly without any further review/feedback.

Reply:
I made some adjustments in CoroCoro which could be used for reference
https://github.com/manga-download/haruneko/blob/master/web/src/engine/websites/CoroCoro.ts#L86

Reply:
Sorry for this misundernstanding.